File: src/cavebot/navigation/walker.py
"""Convert A* paths to arrow key sequences and execute them."""

from __future__ import annotations

import enum
import time

from cavebot.core.input import press_keys_simultaneous


class Direction(enum.Enum):
    NORTH = (0, -1)
    SOUTH = (0, 1)
    WEST = (-1, 0)
    EAST = (1, 0)
    NORTHWEST = (-1, -1)
    NORTHEAST = (1, -1)
    SOUTHWEST = (-1, 1)
    SOUTHEAST = (1, 1)

    @property
    def keys(self) -> list[str]:
        dx, dy = self.value
        result = []
        if dy < 0:
            result.append("up")
        elif dy > 0:
            result.append("down")
        if dx < 0:
            result.append("left")
        elif dx > 0:
            result.append("right")
        return result


_DELTA_TO_DIR = {d.value: d for d in Direction}


def path_to_directions(path: list[tuple[int, int]]) -> list[Direction]:
    directions = []
    for i in range(1, len(path)):
        dx = path[i][0] - path[i - 1][0]
        dy = path[i][1] - path[i - 1][1]
        dx = max(-1, min(1, dx))
        dy = max(-1, min(1, dy))
        direction = _DELTA_TO_DIR.get((dx, dy))
        if direction is not None:
            directions.append(direction)
    return directions


def walk_path(path: list[tuple[int, int]], step_delay: float = 0.2) -> None:
    directions = path_to_directions(path)
    for direction in directions:
        press_keys_simultaneous(direction.keys)
        time.sleep(step_delay)

File: go-cavebot/internal/navigation/walker.go
package navigation

import "time"

type Direction int

const (
	North Direction = iota
	South
	West
	East
	Northwest
	Northeast
	Southwest
	Southeast
)

type dirDelta struct {
	dx, dy int
}

var dirDeltas = map[Direction]dirDelta{
	North:     {0, -1},
	South:     {0, 1},
	West:      {-1, 0},
	East:      {1, 0},
	Northwest: {-1, -1},
	Northeast: {1, -1},
	Southwest: {-1, 1},
	Southeast: {1, 1},
}

var deltaToDir map[dirDelta]Direction

func init() {
	deltaToDir = make(map[dirDelta]Direction)
	for d, delta := range dirDeltas {
		deltaToDir[delta] = d
	}
}

func (d Direction) Keys() []string {
	delta := dirDeltas[d]
	var keys []string
	if delta.dy < 0 {
		keys = append(keys, "up")
	} else if delta.dy > 0 {
		keys = append(keys, "down")
	}
	if delta.dx < 0 {
		keys = append(keys, "left")
	} else if delta.dx > 0 {
		keys = append(keys, "right")
	}
	return keys
}

func PathToDirections(path [][2]int) []Direction {
	var dirs []Direction
	for i := 1; i < len(path); i++ {
		dx := path[i][0] - path[i-1][0]
		dy := path[i][1] - path[i-1][1]
		if dx < -1 {
			dx = -1
		} else if dx > 1 {
			dx = 1
		}
		if dy < -1 {
			dy = -1
		} else if dy > 1 {
			dy = 1
		}
		if d, ok := deltaToDir[dirDelta{dx, dy}]; ok {
			dirs = append(dirs, d)
		}
	}
	return dirs
}

func WalkPath(path [][2]int, stepDelay time.Duration, pressFunc func([]string)) {
	dirs := PathToDirections(path)
	for _, d := range dirs {
		pressFunc(d.Keys())
		time.Sleep(stepDelay)
	}
}
